UI documents tree

falken10vdl opened the pull request:
This is  a proposal to improve workflow of documents in ifc files by:

- Using a tree like structure to edit/add documents in the Documents panel
- Adding an object list that helps with assignation of documents and objects.
![image](https://github.com/user-attachments/assets/ce422554-0693-4ef9-84ae-316a2e1aebb4)
![image](https://github.com/user-attachments/assets/22d8853f-421a-4a62-9e27-2cdf67f22796)

- Assign a context menu for right click when an object is selected in order to see (and launch) documents assinged to the object
![image](https://github.com/user-attachments/assets/74c41326-2b5c-40a6-a995-5d7de8e3e295)

Cheers!

falken10vdl committed:
Implemented tree like structure for documents

falken10vdl committed:
updated document ui

maxfb87 commented:
I like it! Expecially the right click menu and the assigned objects part!

Andrej730 commented:
Thanks! Tree view will be much more easier to use.

I half-read through the PR code, here are the notes I've got so far:

1. Please rebase PR to the latest commit, there seems to be some noise in `model/decorator.py` and `project/decorator.py` unrelated to this PR.

2. `ci-black-formatting` is failing.

3. `ExpandedDocuments` - new property group is unnecessary, please add it to existing document props.

4. `core/document.py` - methods in core are supposed to be very abstract, they don't access ifc or blender entities, they basically just call other methods from `tool/document.py`. So this module should be refactored, otherwise core tests will break.

5. `document/__init__.py` - for consistency please use `bpy.types.VIEW3D_MT_object_context_menu.append` instead of importing it from `bpy.types`

6. Main UI shows total count not just for IfcDocumentInformations, but also for IfcDocumentReferences and referenced objects? Is this neccessary information for the general documents UI? Seems a bit overloaded.

7. `props.documents and props.active_document_index < len(props.documents)` used lots of times - please use `props.active_document` instead.

8. `draw_attributes` - why filtering "Name" attribute?

9. `ObjectDocumentData.documents` - `rel.RelatingDocument` is repeated 21 times, please store it in some variable to optimize how many times we access IFC for this.

10. `hasattr(ref, "ReferencedDocument")` - unnecessary check, IfcDocumentReference always has this attribute in ifc4+.

falken10vdl commented:
@Andrej730 Excelent feedback! Thanks a lot for your comment. In fact I was doing some other PR in the cost area and got very consistent feedback from @maxfb87 . I am still quite new to the bonsai (and blender) architecture so these comments are greatly appreciated!
If you do not mind I will sync the fork I have in https://github.com/falken10vdl/IfcOpenShell/tree/v0.8.0 and generate a clean PR based on all the info you have provided. If you recall I already tried to rebase and for some unknown reason to me instead of applying my changes to the new base, it applied all the commits from the old base to the new one + my commits. Still have to do some testing on why I cannot get it working properly so I rather just start a clean one to push if you do not mind.

Andrej730 commented:
@falken10vdl can you please try to rebase this PR? This issue will keep coming back. Just to keep the entire PR history connected in one thread instead of producing new PRs for each iteration.

maxfb87 commented:
@Andrej730 hi, can i ask you what does it mean "ci-black-formatting is failing"?

Edit: i found it and now i understand but i have another question: is it true that if i run "black ." from the repo it does use the .toml configuration file already in the repo?

falken10vdl commented:
@Andrej730 I will try... Hopefully the Rebase is not proposing tens of commits done by other members... I will do some tests before.

Andrej730 commented:
> Edit: i found it and now i understand but i have another question: is it true that if i run "black ." from the repo it does use the .toml configuration file already in the repo?

Yes, `black .` is a go-to way to apply black formatting to all files using settings from .toml. 
It is the way this github workflow does it:
https://github.com/IfcOpenShell/IfcOpenShell/blob/d7cf803fdfb38df1e3eb37879cc04f733339b704/.github/workflows/ci-black-formatting.yaml#L40-L44

falken10vdl committed:
Refactoring based on developer's feedback

falken10vdl commented:
@Andrej730 
Here the updates done:

>    Please rebase PR to the latest commit, there seems to be some noise in model/decorator.py and project/decorator.py unrelated to this PR.

done. I think I start to know the workflow to not mess up things...

>    ci-black-formatting is failing.

black formatting applied

>    ExpandedDocuments - new property group is unnecessary, please add it to existing document props.

done

>    core/document.py - methods in core are supposed to be very abstract, they don't access ifc or blender entities, they basically just call other methods from tool/document.py. So this module should be refactored, otherwise core tests will break.

Please take a look to the new setup of methods in core and tool. Hopefully I have followed better the design philosophy

>    document/__init__.py - for consistency please use bpy.types.VIEW3D_MT_object_context_menu.append instead of importing it from bpy.types

done

>    Main UI shows total count not just for IfcDocumentInformations, but also for IfcDocumentReferences and referenced objects? Is this neccessary information for the general documents UI? Seems a bit overloaded.

Just tell me what you guys think is good to have there. i agree I always like to put lots of infos... but clarity is hit

>    props.documents and props.active_document_index < len(props.documents) used lots of times - please use props.active_document instead.

done

>    draw_attributes - why filtering "Name" attribute?

I understood that name is optional for IfcDocumentReference (https://standards.buildingsmart.org/IFC/DEV/IFC4_2/FINAL/HTML/schema/ifcexternalreferenceresource/lexical/ifcdocumentreference.htm). In the previous code I think that it was not initialized when a docrefernce was created so then editing afterwards was confusing as the reference + description was used for UI and sorting

>    ObjectDocumentData.documents - rel.RelatingDocument is repeated 21 times, please store it in some variable to optimize how many times we access IFC for this.

done

>    hasattr(ref, "ReferencedDocument") - unnecessary check, IfcDocumentReference always has this attribute in ifc4+.

removed

Again thanks a lot @Andrej730 for such great feedback!

Andrej730 commented:
> Here the updates done:
> 
> > Please rebase PR to the latest commit, there seems to be some noise in model/decorator.py and project/decorator.py unrelated to this PR.
> 
> done. I think I start to know the workflow to not mess up things...

Great, glad rebase thing worked out.


> > core/document.py - methods in core are supposed to be very abstract, they don't access ifc or blender entities, they basically just call other methods from tool/document.py. So this module should be refactored, otherwise core tests will break.
> 
> Please take a look to the new setup of methods in core and tool. Hopefully I have followed better the design philosophy

That's much more clean. But there's a lot of tests in document module now failing - please check.

> > Main UI shows total count not just for IfcDocumentInformations, but also for IfcDocumentReferences and referenced objects? Is this neccessary information for the general documents UI? Seems a bit overloaded.
> 
> Just tell me what you guys think is good to have there. i agree I always like to put lots of infos... but clarity is hit

I mean showing total count of objects in UI list is a common pattern in Bonsai UI, but more than that is usually a noise that's rarely needed.

> > props.documents and props.active_document_index < len(props.documents) used lots of times - please use props.active_document instead.
>
> done

There's still one left in prop.py.

> > draw_attributes - why filtering "Name" attribute?
> 
> I understood that name is optional for IfcDocumentReference ([standards.buildingsmart.org/IFC/DEV/IFC4_2/FINAL/HTML/schema/ifcexternalreferenceresource/lexical/ifcdocumentreference.htm](https://standards.buildingsmart.org/IFC/DEV/IFC4_2/FINAL/HTML/schema/ifcexternalreferenceresource/lexical/ifcdocumentreference.htm)). In the previous code I think that it was not initialized when a docrefernce was created so then editing afterwards was confusing as the reference + description was used for UI and sorting

At first I though preventing Name from being edited seems arbitrary, but now I see where it's coming from, I've impemented it myself fe316e2 🙃

But still, any particular error you've met? What's the point of skipping it in `draw_attributes` if it's already skipped during import?

Also, useful comments in `import_document_attributes` were removed, now it looks very abitrary for someone reading just the code.
 
> > hasattr(ref, "ReferencedDocument") - unnecessary check, IfcDocumentReference always has this attribute in ifc4+.
> done

still present in document.py


I was able to read deeper into the code this time, here are the new notes:
1. `json` imported to document.py twice
2. `bim.open_ifc_document` - it's unreliable to expect `blender` available from PATH. Current executable should be used for opening new IFC file.
3. `bim.toggle_document` - is `tool.Ifc.Operator` needed here? It's only needed when there are changes to IFC.
4. `load_document_objects_into_props` - why check `"document_objects" not in cls.data`? Shouldn't it just load data if it's not loaded?
5. `load_document_objects_into_props` - please use `natsorted`
6. `update_document_objects` - I think this should be moved to `tool` and should be called from corresponding methods - e.g. `core.load_project_documents`, `core.load_document`, etc. And maybe it should just reset `active_document_index` to trigger the callback to keep it simple? Seems to do the same thing.
7. `bim.load_document` operator seems useless now.
8. `bim.disable_object_document_editing_ui`, `bim.enable_editing_document`, `bim.disable_editing_document` - it seems execute code should be moved to `core`
9. `enable_editing_ui` / `disable_editing_ui` do the same as `enable_document_editing` / `disable_document_editing`? There is a bug when you disable document editing, the entire Documents UI is disabled.
10. `bim.update_assigned_documents` - is there any point of having it as operator if it's not exposed anywhere? Shouldn't it be a method in tool? And there is something odd is going on with this method and `bim.edit_document` - lots of data resets. Is there a way to simplify this? Btw, `bonsai.bim.handler.refresh_ui_data` happens automatically at the end of all `tool.Ifc.Operator`.
11. `bim.add_information` - it's safe to assume that there is an IfcProject, otherwise file wouldn't load to Bonsai 😁 It seems there's some special convention to use -1 for IfcProject. Maybe there should be some kind of enum property on `Document` in prop that would clearly state what this item is - PROJECT, INFORMATION, REFERENCE? It would make things much more readable.
And I don't understand what the point of `virtual_root_id` in `expanded_documents` - is "Project Documents" folder suppose to be collapsable too?
12. `bim.add_document_reference`, `bim.add_information`, `bim.assign_document`, `bim.unassign_document` - perhaps code should be moved to tool/core? And some checks doesn't seem necessary as those operators won't be available from UI if those checks do not pass. And is there any reason for checking whether `json.loads` failed - any cases when it might get invalid?
13. `props.is_document_editing` - isn't it the same as `active_document_id`?
14. `bim.load_object_documents` - duplicated code for filling `assigned_documents`. We don't usually use data from data.py in our operators - it's for UI only, if needed we access IFC directly.

falken10vdl committed:
updates based on developers feedback

falken10vdl committed:
misc panel working

falken10vdl committed:
nicer ui

falken10vdl committed:
cleanup and formatting

falken10vdl commented:
> 1. json imported to document.py twice

done

> 2. bim.open_ifc_document - it's unreliable to expect blender available from PATH. Current executable should be used for opening new IFC file.

done

> 3. bim.toggle_document - is tool.Ifc.Operator needed here? It's only needed when there are changes to IFC.

done

> 4. load_document_objects_into_props - why check "document_objects" not in cls.data? Shouldn't it just load data if it's not loaded?

done

> 5. load_document_objects_into_props - please use natsorted

done

> 6. update_document_objects - I think this should be moved to tool and should be called from corresponding methods - e.g. core.load_project_documents, core.load_document, etc. And maybe it should just reset active_document_index to trigger the callback to keep it simple? Seems to do the same thing.
moved to tool/document.py

> 7. bim.load_document operator seems useless now.

removed

> 8. bim.disable_object_document_editing_ui, bim.enable_editing_document, bim.disable_editing_document - it seems execute code should be moved to core

I have moved things to core but still my knowledge on what to put in core, tool or data is very imperfect

> 9. enable_editing_ui / disable_editing_ui do the same as enable_document_editing / disable_document_editing? There is a bug when you disable document editing, the entire Documents UI is disabled.

Bug fixed and logic simplified

> 10. bim.update_assigned_documents - is there any point of having it as operator if it's not exposed anywhere? Shouldn't it be a method in tool? And there is something odd is going on with this method and bim.edit_document - lots of data resets. Is there a way to simplify this? Btw, bonsai.bim.handler.refresh_ui_data happens automatically at the end of all tool.Ifc.Operator.

This is an interesting one. I could not find a way to update a UIList for assigned documents (in object information misc tab ) without invoking an operator. It refused to do it being in the ui part... In the end triggering an update of the UIList when object selection was changed was very cumbersome (or I did not find a simple way) So I just simplified the whole logic removing that UIList for assigned documents

> 11. bim.add_information - it's safe to assume that there is an IfcProject, otherwise file wouldn't load to Bonsai 😁 It seems there's some special convention to use -1 for IfcProject. Maybe there should be some kind of enum property on Document in prop that would clearly state what this item is - PROJECT, INFORMATION, REFERENCE? It would make things much more readable. And I don't understand what the point of virtual_root_id in expanded_documents - is "Project Documents" folder suppose to be collapsable too?

Yes, I agree it is alittle bit absurd to make collapsable "projec documents". your suggestion to use an ENUM is much better

> 12. bim.add_document_reference, bim.add_information, bim.assign_document, bim.unassign_document - perhaps code should be moved to tool/core? And some checks doesn't seem necessary as those operators won't be available from UI if those checks do not pass. And is there any reason for checking whether json.loads failed - any cases when it might get invalid?

This is where I feel less confident. Still trying to understand what makes sense to move and what not...

> 13. props.is_document_editing - isn't it the same as active_document_id?

In a way yes... I juts found it like easier to read but I agree that active_document_id == 0 is equivalent to  props.is_document_editing == False  and active_document_id > 0 equivalent to props.is_document_editing == True

> 14. bim.load_object_documents - duplicated code for filling assigned_documents. We don't usually use data from data.py in our operators - it's for UI only, if needed we access IFC directly.

I removed all related to assigned_documents

Andrej730 commented:
@falken10vdl 

> That's much more clean. But there's a lot of tests in document module now failing - please check.

1. Tests are still failing.

> > Just tell me what you guys think is good to have there. i agree I always like to put lots of infos... but clarity is hit
> 
> I mean showing total count of objects in UI list is a common pattern in Bonsai UI, but more than that is usually a noise that's rarely needed.

2. Nice, there's no more separate reference count. But do we still need total referenced object count? Seems like a very abstract thing - you typically would need this information only for specific documents, not for the entire project.

3. 
> But still, any particular error you've met? What's the point of skipping it in draw_attributes if it's already skipped during import?

`draw_attributes` still have unused `filter_attributes`.

4.
> >  json imported to document.py twice
> 
> done

Why `bonsai.bim.helper` import is removed?

5.  > `from .data import DocumentData, ObjectDocumentData`

Please don't use relative imports for consistency.

6. > > 3\. bim.open_ifc_document - it's unreliable to expect blender available from PATH. Current executable should be used for opening new IFC file.
> 
> done

Is there a need to check exception here?

7. > > bim.disable_object_document_editing_ui, bim.enable_editing_document, bim.disable_editing_document - it seems execute code should be moved to core
> 
> I have moved things to core but still my knowledge on what to put in core, tool or data is very imperfect

There should be no direct attributes access in core (as currently there is in `disable_object_document_editing_ui`), just tool method calls.

There's also now some unused `props` in core.

> > . bim.add_document_reference, bim.add_information, bim.assign_document, bim.unassign_document - perhaps code should be moved to tool/core? And some checks doesn't seem necessary as those operators won't be available from UI if those checks do not pass. And is there any reason for checking whether json.loads failed - any cases when it might get invalid?
> 
> This is where I feel less confident. Still trying to understand what makes sense to move and what not...

In theory there should be just 1 core method called by the operator and everything else is moved to tool (especially some simple methods that just edit Blender props). I guess apart from operator error handling.

Also, some operators seem to invalidate data using `is_loaded = False`. IFC Operators will do it automatically. If it's not an IFC Operator, then why reload data if no ifc was changed?

falken10vdl committed:
updates based on core developer's feedback

falken10vdl commented:
> @falken10vdl
> 
> > That's much more clean. But there's a lot of tests in document module now failing - please check.
> 
>     1. Tests are still failing.

Yes. I have not yet taken the time to go on how the testing systems works. I need to find them and run them. This is described here right? https://docs.bonsaibim.org/guides/development/running_tests.html

> 
> > > Just tell me what you guys think is good to have there. i agree I always like to put lots of infos... but clarity is hit
> > 
> > 
> > I mean showing total count of objects in UI list is a common pattern in Bonsai UI, but more than that is usually a noise that's rarely needed.
> 
>     2. Nice, there's no more separate reference count. But do we still need total referenced object count? Seems like a very abstract thing - you typically would need this information only for specific documents, not for the entire project.

Removed

> 
> > But still, any particular error you've met? What's the point of skipping it in draw_attributes if it's already skipped during import?
> 
> `draw_attributes` still have unused `filter_attributes`.
> 

removed that option

> > > json imported to document.py twice
> > 
> > 
> > done
> 
> Why `bonsai.bim.helper` import is removed?

Sorry. It was my bad. I use pylance extension for vscode and it wrongly thinks that bonsai.bim.helper is not used. It thinks that bonsai is used but since the import is bonsai.bim.helper, unless we change it to an import using from or alias it will not recognize it. 
 
>     5. > `from .data import DocumentData, ObjectDocumentData`
> 
> 
> Please don't use relative imports for consistency.

Done

>     6. > > 3. bim.open_ifc_document - it's unreliable to expect blender available from PATH. Current executable should be used for opening new IFC file.
> 
> 
> > done
> 
> Is there a need to check exception here?

Removed. I got sometime in linux an error when spanning out processes while doing other activities in teh background but I agree this would be a rather weird case

>     7. > > bim.disable_object_document_editing_ui, bim.enable_editing_document, bim.disable_editing_document - it seems execute code should be moved to core
> 
> 
> > I have moved things to core but still my knowledge on what to put in core, tool or data is very imperfect
> 
> There should be no direct attributes access in core (as currently there is in `disable_object_document_editing_ui`), just tool method calls.
> 
> There's also now some unused `props` in core.


I have made some updates. Hopefully it is cleaner now

> > > . bim.add_document_reference, bim.add_information, bim.assign_document, bim.unassign_document - perhaps code should be moved to tool/core? And some checks doesn't seem necessary as those operators won't be available from UI if those checks do not pass. And is there any reason for checking whether json.loads failed - any cases when it might get invalid?
> > 
> > 
> > This is where I feel less confident. Still trying to understand what makes sense to move and what not...
> 
> In theory there should be just 1 core method called by the operator and everything else is moved to tool (especially some simple methods that just edit Blender props). I guess apart from operator error handling.
> 
> Also, some operators seem to invalidate data using `is_loaded = False`. IFC Operators will do it automatically. If it's not an IFC Operator, then why reload data if no ifc was changed?

I have changed it to follow the same logic as in module "group"

Thanks @Andrej730 !

falken10vdl committed:
cleanup nomenclature and some redundant code

Andrej730 commented:
> Yes. I have not yet taken the time to go on how the testing systems works. I need to find them and run them. This is described here right? [docs.bonsaibim.org/guides/development/running_tests.html](https://docs.bonsaibim.org/guides/development/running_tests.html)

Yes.

falken10vdl committed:
updated to get make test-tool MODULE=document working

falken10vdl committed:
updated to get make test-bim MODULE=document working

falken10vdl commented:
@Andrej730 Please can you take a look? I have updated the tests for test-tool and test-bim. Are there other tests that need to be run for the documents module?
Thanks!

falken10vdl committed:
adapted to get  pytest -p no:pytest-blender test/core/test_document.py working. Black formating

falken10vdl commented:
Also updated for 

```
falken10vdl@monster:~/bonsaiDevel/IfcOpenShell/src/bonsai$ pytest -p no:pytest-blender test/core/test_document.py
============================================================================================================== test session starts ===============================================================================================================
platform linux -- Python 3.11.13, pytest-8.4.1, pluggy-1.6.0
rootdir: /home/falken10vdl/bonsaiDevel/IfcOpenShell/src/bonsai
configfile: pytest.ini
plugins: blender-3.0.7, bdd-8.1.0
collected 16 items                                                                                                                                                                                                                               

test/core/test_document.py ................                                                                                                                                                                                                [100%]

=============================================================================================================== 16 passed in 0.02s ===============================================================================================================

```

Thanks!

falken10vdl commented:
@Andrej730 are you happy to get this one merged?
Thanks!

falken10vdl commented:
Let me rebase and merge and see people's feedback.
